benchmark/benchmain/main.go

```diff
@@ -365,178 +413,202 @@ func runBenchmark(caller func(int), startTimer func(), stopTimer func(uint64), b
-	flag.StringVar(&memProfile, "memProfile", "", "Enables memory profiling output to the filename provided.")
-	flag.IntVar(&memProfileRate, "memProfileRate", 512*1024, "Configures the memory profiling rate. \n"+
-		"memProfile should be set before setting profile rate. To include every allocated block in the profile, "+
-		"set MemProfileRate to 1. To turn off profiling entirely, set MemProfileRate to 0. 512 * 1024 by default.")
-	flag.StringVar(&cpuProfile, "cpuProfile", "", "Enables CPU profiling output to the filename provided")
-	flag.StringVar(&compressorMode, "compression", modeOff,
-		fmt.Sprintf("Compression mode - One of: %v", strings.Join(allCompressionModes, ", ")))
-	flag.StringVar(&preloaderMode, "preloader", modeOff,
-		fmt.Sprintf("Preloader mode - One of: %v", strings.Join(allPreloaderModes, ", ")))
-	flag.StringVar(&benchmarkResultFile, "resultFile", "", "Save the benchmark result into a binary file")
-	flag.StringVar(&networkMode, "networkMode", "", "Network mode includes LAN, WAN, Local and Longhaul")
+// featureOpts represents options which can have multiple values. The user
+// usually provides a comma-separated list of options for each of these
+// features through command line flags. We generate all possible combinations
+// for the provided values and run the benchmarks for each combination.
+type featureOpts struct {
+	enableTrace        []bool          // Feature index 0
+	readLatencies      []time.Duration // Feature index 1
+	readKbps           []int           // Feature index 2
+	readMTU            []int           // Feature index 3
+	maxConcurrentCalls []int           // Feature index 4
+	reqSizeBytes       []int           // Feature index 5
+	respSizeBytes      []int           // Feature index 6
+	compModes          []string        // Feature index 7
+	enableChannelz     []bool          // Feature index 8
+	enablePreloader    []bool          // Feature index 9
+}
+
+// generateFeatures generates all combinations of the provided feature options.
+// For example, let's say the user sets -workloads=all and
+// -maxConcurrentCalls=1,100, this would end up with the following
+// combinations:
+// [workloads: unary, maxConcurrentCalls=1]
+// [workloads: unary, maxConcurrentCalls=1]
+// [workloads: streaming, maxConcurrentCalls=100]
+// [workloads: streaming, maxConcurrentCalls=100]
+// [workloads: unconstrained, maxConcurrentCalls=1]
+// [workloads: unconstrained, maxConcurrentCalls=100]
+func (b *benchOpts) generateFeatures() []stats.Features {
+	// We use reflection here to get the number of features. Then we get the
+	// number of values for each feature. Each combination is represented using a
+	// slice (size=#features), where each item represents one feature. The value
+	// of each item is an index into a slice which contains all possible values
+	// for that feature. Based off of each combination, we create stats.Feature
+	// objects.
+	value := reflect.ValueOf(*b.features)
```

Comment:
Since the order of the features matter, why we bother with reflection and not use the old way to generate featuresNum?

Reply:
In that case let's move the generation of `featuresNum` into a separate function. Call that function from main and store `featuresNum`.Pass `featuresNum` as a parameter to `generateFeatures` and `sharedFeatures`. By doing this, we can contain the use of reflection _magic_.